HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,349 @@
+  desc.add<edm::ParameterSetDescription>("L2CandSelection", descMuonSelL2);
+
+  edm::ParameterSetDescription descMuonSelL3;
+  MuonSelectionCuts::fillPSetDescription(descMuonSelL3);
+  desc.add<edm::ParameterSetDescription>("L3CandSelection", descMuonSelL3);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+  if (saveTags()) {
+    filterproduct.addCollectionTag(L3candTag_);
+    filterproduct.addCollectionTag(L2CandTag_);
+  }
+  // get hold of trks
+  auto const L3mucands = iEvent.getHandle(L3candToken_);
+  auto const L2mucands = iEvent.getHandle(L2CandToken_);
+  auto const previousLevelCands = iEvent.getHandle(previousCandToken_);
+  std::vector<reco::RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(trigger::TriggerMuon, vcands);
+  }
+
+  auto const& beamSpot = iEvent.get(beamspotToken_);
+
+  // Number of objects passing the L3 Trigger:
+  int nL3 = 0;
+  for (unsigned int iMuL3 = 0; iMuL3 < L3mucands->size(); iMuL3++) {
+    reco::RecoChargedCandidateRef L3cand(L3mucands, iMuL3);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L3cand isNonnull " << L3cand.isNonnull();
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L3cand, vcands))
+      continue;
+    // apply all the cuts
+    if (!applySelection(L3cand, beamSpot, cutParsL3_))
+      continue;
+    nL3++;
+    filterproduct.addObject(trigger::TriggerMuon, L3cand);
+  }
+  // get the L3 muons that passed the selection
+  std::vector<reco::RecoChargedCandidateRef> passingL3mucands;
+  filterproduct.getObjects(trigger::TriggerMuon, passingL3mucands);
+
+  // Number of objects passing the L2 Trigger:
+  int nL2 = 0;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    reco::RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "L2cand isNonnull " << L2cand.isNonnull();
+    // did this candidate triggered at previous stage.
+    if (matchPreviousCand_ && !triggeredByPreviousLevel(L2cand, vcands))
+      continue;
+    // check if L2 and L3 candidates are the same
+    if (checkOverlap(L2cand, passingL3mucands, maxDeltaR2_L2L3_)) {
+      continue;
+    }
+    // apply all the cuts
+    if (!applySelection(L2cand, beamSpot, cutParsL2_))
+      continue;
+    nL2++;
+    filterproduct.addObject(trigger::TriggerMuon, L2cand);
+  }
+
+  // filter decision
+  const bool acceptL3(nL3 >= cutParsL3_.min_N);
+  const bool acceptL2(nL2 >= cutParsL2_.min_N);
```

Comment:
If I remember correctly, the final filter result was based on the sum of L3 and L2 (L2+L3 > min_N). I thought this was the reason to name the plugin "L3orL2". Now this changes to two requirements (L2 > minN_L2 and L3 > minN_L3). I just want to ask if this is intended, and if the name of the plugin should change based on this.

Reply:
Yes I think we should change the name, but in the past it wasn't simply L2+L3, it saved all the L3 and all the L2 that didn't match to any L3. Now it does the same but also controlling the number of L2 and L3 independently.